meteor/server/databaseMigrationSystemSteps.ts

```diff
@@ -102,6 +102,127 @@ function ensureCollectionProperty<T = any> (
 		}
 	}
 }
+/**
+ * Convenience function to generate basic test
+ * øparam type
+ */
+function ensurePlayoutDevicesHostAndPort (
+	deviceType: PlayoutDeviceType = -1
+): MigrationStepBase | null {
+	let collection: Mongo.Collection<any> = Collections.PeripheralDevices
+	if (!collection) throw new Meteor.Error(404, `Collection PeripheralDevices not found`)
+
+	let allDevices = {}
+
+	// iterate over playout gateways
+	collection.find({'type': PeripheralDeviceAPI.DeviceType.PLAYOUT}).fetch().forEach((playoutGw) => {
+		let devices: PlayoutDeviceSettings = objectPathGet(playoutGw, 'settings.devices', {})
+		// filter on device-type unless devicetype === -1
+		if (deviceType > -1) {
+			devices = _.pick(devices, (device) => device.type === deviceType)
+		}
+		allDevices[playoutGw._id] = devices
+	})
+
+	// iterate over devices for gateways
+	_.each(allDevices, (devices, key) => {
+		allDevices[key] = _.pick(devices, (device) => {
+			let needsMoreData = false
+			let foundHostProp = objectPathGet(device, 'options.host', undefined) 
+			let foundPortProp = objectPathGet(device, 'options.port', undefined) 
+			// let foundHostProp = objectPathGet(device, 'options.host', '') // @ TODOD: discuss if host should be required to be initiated as empty string
+			if (foundHostProp !== undefined && typeof foundHostProp === 'string') {
+				// if host is string but length is 0, then add to mossingHostDevices
+				if(foundHostProp.trim().length === 0){
+					needsMoreData = true
+				}
+			}
+
+			// port not number or <= 0
+			if(foundPortProp !== undefined && (isNaN(foundPortProp) || foundPortProp <= 0)) {
+				needsMoreData = true
+			}
+			
+			return needsMoreData
+		})
+	})	
+
+	// create migrationsteps for all affected devices
+	return {
+		id: `ensureHostPortOnDevicesOfType${deviceType > -1 ? deviceType : 'All'}`,
+		canBeRunAutomatically: false,
+		validate: () => {
+			for (let key in allDevices) {
+				if (!_.isEmpty(allDevices[key])) {
+					return 'Some playout devices are missing host or port'
+				}
+			}
+			return true
+		},
+		input: () => {
+			// let objects = collection.find(selector).fetch()
+			let inputs: Array<MigrationStepInput> = []
+			
+			_.each(allDevices, (devices: {[key:string]: PlayoutDeviceSettingsDevice}, gatewayId: string) => {
+				_.each(devices, (device: PlayoutDeviceSettingsDevice, deviceId: string) => {
+
+					// should have host added
+					let foundHostProp = objectPathGet(device, 'options.host', undefined) 
```

Comment:
@nytamin Repeating myself here, but I don't see better patterns as we have this nested structure with potentially multiple gateways fucking up better structures. And: repeating things twice is not an anti pattern to me ;)